pennylane/gradients/gradient_transform.py

```diff
@@ -144,8 +164,32 @@ def jacobian_wrapper(*args, **kwargs):
                 # is present inside the QNode.
                 return qjac
 
-            # Classical processing of a single argument is present. Return qjac @ cjac.
-            jac = qml.math.squeeze(qml.math.tensordot(qml.math.T(cjac), qjac, [[-1], [-1]]))
-            return qml.math.T(jac)
+            # Classical processing present of either:
+            #   a) a single argument
+            #   b) multiple arguments of the same shape
+            # The shape of the classical jacobian returned by qml.jacobian depends on the scenario:
+            #   a) (# gate args, qnode arg shape)
+            #   b) (reverse qnode arg shape, # gate args, # qnode args)
+            # It then needs to be contracted with the quantum jacobian of shape:
+            #      (qnode output shape, # gate args)
+            # The result should have shape:
+            #   a) (qnode ouput shape, qnode arg shape)
+            #   b) (reverse qnode arg shape, qnode output shape, # qnode args)
+            num_gate_args = qml.math.shape(qjac)[-1]
+            trainable_args_idx = self._jacobian_trainable_args(args, qnode.interface)
+            qnode_arg_shape = qml.math.shape(args[trainable_args_idx[0]])

```

Comment:
Maybe good to leave a comment here (lmk if I misunderstood what's going on here).
```suggestion
            # Either a single argument or multiple with same shape; we can obtain the
            # shape of the first trainable QNode argument
            qnode_arg_shape = qml.math.shape(args[trainable_args_idx[0]])
```